src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -437,29 +523,109 @@ HeapWord* ShenandoahFreeSet::try_allocate_in(ShenandoahHeapRegion* r, Shenandoah
     // available free segments within regions for more efficient searches for "good fit".
 
     // Record the remainder as allocation waste
+    size_t idx = r->index();
     if (req.is_mutator_alloc()) {
       size_t waste = r->free();
       if (waste > 0) {
         increase_used(waste);
         generation->increase_allocated(waste);
         _heap->notify_mutator_alloc_words(waste >> LogHeapWordSize, true);
       }
-    }
-
-    size_t num = r->index();
-    _collector_free_bitmap.clear_bit(num);
-    _mutator_free_bitmap.clear_bit(num);
-    // Touched the bounds? Need to update:
-    if (touches_bounds(num)) {
-      adjust_bounds();
+      assert(probe_mutator_set(idx), "Must be mutator free: " SIZE_FORMAT, idx);
+      remove_from_mutator_set(idx);
+      assert(!in_collector_set(idx) && !in_old_collector_set(idx), "Region cannot be in multiple free sets");
+      adjust_mutator_bounds_if_touched(idx);
+    } else if (r->free() < PLAB::min_size() * HeapWordSize) {
+      // Permanently retire this region if there's room for a fill object
+      size_t waste = r->free();
```

Comment:
Are we writing a filler object here to keep this region out of the old collector freeset when it's rebuilt?

Reply:
yes.  Putting the region back into the old_collector set at rebuild time just makes work for us, because then we'll have to rediscover that it needs to be retired, and will need to retire it, and will need to adjust the bounds again.

Reply:
I'm adding a comment to the code to clarify.

Reply:
Should we have a more restrictive test when rebuilding the freeset? Rather than including regions with non-zero allocation capacity, include only regions with capacity greater than some minimum? (Minimum LAB size, e.g.)

Reply:
In the very latest revision committed, I have removed the fill object and have taken your suggestion to only install regions into a freeset if available memory is greater than PLAB::min_size().  Thanks.  I think this is cleaner.